torch/csrc/nnpack/NNPACK.cpp

```diff
@@ -0,0 +1,348 @@
+#include "NNPACK.h"
+
+#include "TH/TH.h"
+#include <stdlib.h>
+
+namespace torch {
+namespace nnpack {
+
+// Stolen from Caffe2
+static pthreadpool_t nnpack_threadpool_ = nullptr;
+
+pthreadpool_t nnpack_threadpool() {
+  if (nnpack_threadpool_ == nullptr) {
+    enum nnp_status nnpack_status = nnp_initialize();
+    if (nnpack_status != nnp_status_success) throw std::runtime_error("could not initialize NNPack");
+    nnpack_threadpool_ = pthreadpool_create(16);
+  }
+  return nnpack_threadpool_;
+}
+
+// Assuming for now this is Thread-safe, but that seems like a stretch
+static void *workspace = nullptr;
+static size_t workspace_size = 0;
+
+// NNPack has alignment requirements
+const size_t nnpack_memory_alignment_boundary = 64;
+
+static inline void deallocate_workspace() {
+  if (workspace)
+    std::free(workspace);
+  workspace = nullptr;
+}
+
+static inline void allocate_workspace() {
+  if (workspace)
+    deallocate_workspace();
+  posix_memalign(&workspace, nnpack_memory_alignment_boundary, workspace_size);
+}
+
+void SpatialConvolution_updateOutput(
+    at::Tensor& input,
+    at::Tensor& output,
+    at::Tensor& weight,
+    at::Tensor& bias,
+    int kW,
+    int kH,
+    int padW,
+    int padH) {
+  // Setup parameters for the NNPack convolution output function call
+
+  // For now, we use the default algorithm
+  auto algorithm = nnp_convolution_algorithm_auto;
+
+  // Our input Tensor must be in the form N,C,H,W
+  if (input.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D input Tensor N,C,H,W");
+  }
+  // Our weight Tensor must be in the form oC,iC,kH,kW
+  if (weight.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D weight Tensor oC,iC,kH,kW");
+  }
+  // Our weight Tensor must be in the form N,oC,oH,oW
+  if (output.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D output Tensor N,oC,oH,oW");
+  }
```

Comment:
Does nnpack have size checks? If not, we should check them ourselves

Reply:
Nope, I can add more detailed checks now that we are cool with this overall approach.